### PR TITLE
Fix CircleCi style checker

### PR DESCRIPTION
**Pull request by wilzbach**
It seems like we/I accidentally disabled the style target on CircleCi (`style` is not defined in the command switch). I take the full blame for this, but at least it's a good proof that the CI actually does catch quite a few things ;-)

I split up the fixes depending on their category.

There's only [one line](https://github.com/dlang/phobos/blob/master/std/regex/package.d#L421) left about which DScanner is complaining (hence the "blocked" label):

```
public static immutable ctRegex(alias pattern, alias flags=[]) = ctRegexImpl!(pattern, flags).nr;
```

It warns with:

```
std/regex/package.d(421:32)[error]: Declaration expected
std/regex/package.d(421:62)[error]: Declaration expected
std/regex/package.d(421:64)[error]: Declaration expected
std/regex/package.d(421:97)[warn]: Empty declaration
```

@Hackerpilot is there a simple fix? Otherwise I will add this file to the Dscanner blacklist.

### CODE REVIEW

**Comment by Hackerpilot** [comment]
I'll have to figure out what kind of declaration that is and then update the parser.

**Comment by Hackerpilot** [comment]
This should help: https://github.com/Hackerpilot/Dscanner/releases/tag/v0.4.0-beta.1

**Comment by Hackerpilot** [comment]
Interesting. The compiler used by Travis can't find the `mallocator` module.

**Comment by MartinNowak** [comment]
> Why is it using dmd-2.068.2?

Because you're not specifying with which dmd to build dscanner. It just picks up whatever system-wide dmd installed. Please fix that, e.g. similarly to how we pin stable dmd versions for building dlang.org's ddocs or dmd's bootstrap compiler.

**Comment by wilzbach** [comment]
... and now we run into problem that currently our installer is broken for point releases :/
(See https://github.com/dlang/installer/pull/201 for more details)

**Comment by wilzbach** [comment]
Okay I once again updated this PR to include the changes that managed to sneak in over the last month.
I also had to disable Dscanner's new check for _auto return functions_ which return `void` as it's commonly used like `version(Stddoc) auto myfunction(){ assert(0)}` or with `mixin`.

I hope that this version now passes our CI, s.t. we can finally enable the check again :)

File: std/random.d [hunk 1]
```diff
@@ -1877,7 +1877,8 @@ auto ref choice(Range, RandomGen = Random)(auto ref Range range,
 
 unittest
 {
-    import std.algorithm : canFind, map;
+    import std.algorithm.searching : canFind;
+    import std.algorith.iteration : map;
```
Comment (JackStouffer) [inline std/random.d]: Failing on this line. `algorithm` is misspelled

**Comment by wilzbach** [comment]
I went through another round of catching up & thus rebased to the lastest changes. It would be quite nice if we can get this in rather sooner that later :)

File: std/experimental/ndslice/internal.d [hunk 2]
```diff
@@ -92,13 +92,13 @@ struct PtrShell(Range)
         mixin (`_shift ` ~ op ~ `= shift;`);
     }
 
-    auto opBinary(string op)(sizediff_t shift)
+    typeof(this) opBinary(string op)(sizediff_t shift)
```
Comment (CyberShadow) [inline std/experimental/ndslice/internal.d]: This is technically a false positive, right?

**Comment by CyberShadow** [comment]
Auto-merge toggled on

**Comment by CyberShadow** [comment]
How come I don't see CircleCI in the CI services list on this PR?

**Comment by CyberShadow** [comment]
Auto-merge toggled off

**Comment by JackStouffer** [comment]
It's actually disappeared from all PRs, so probably not due to these changes.

**Comment by CyberShadow** [comment]
Perhaps it's because of this setting?



Link goes to https://circleci.com/docs/fork-pr-builds/.

I don't know how CircleCI is configured (heck, I have no clue why we even need it at all considering there's Travis and AppVeyor), so I'm not sure if it's safe to enable or if it's relevant at all. @wilzbach ?

**Comment by wilzbach** [comment]
> How come I don't see CircleCI in the CI services list on this PR?

Oh sorry - I was hoping that it was just a temporary outage and would start to build the PR.
-> I will have a look :(

> I don't know how CircleCI is configured (heck, I have no clue why we even need it at all considering there's Travis and AppVeyor), so I'm not sure if it's safe to enable or if it's relevant at all. @wilzbach ?

The only reason for using CircleCi is because it's not Travis :P 
With the free plan an organization (e.g. `dlang`) gets five concurrent slots and well with all it's parallel builds `dmd` uses all of them.

**Comment by CyberShadow** [comment]
> The only reason for using CircleCi is because it's not Travis :P

So, to clarify... CircleCI has more throughput than Travis, or using both gets us more throughput than using just one, or something else?

**Comment by wilzbach** [comment]
> So, to clarify... CircleCI has more throughput than Travis, or using both gets us more throughput than using just one, or something else?

Yes it's solely about having more throughput. We tried to use Travis for Phobos before and the builds were stuck in the queue for days, hence the move to CircleCi.
The use of CircleCi for code coverage testing at `dmd` and `druntime` has a similar motivation.

I don't know the reason for the outage with CircleCi yet, but unfortunately it doesn't have anything to do with the "Permissive build setting").
Anyways thanks to @MartinNowak the CI logic is in this separate [`circleci.sh`](https://github.com/dlang/phobos/blob/master/circleci.sh), so we are pretty agnostic about the CI provider.

File: std/experimental/ndslice/internal.d (hunk 2 above)
Comment (wilzbach) [inline std/experimental/ndslice/internal.d]: Yes unfortunately Dscanner can't expand mixins atm, s.t. is the poor man's workaround "for the  greater good".

**Comment by wilzbach** [comment]
> Perhaps it's because of this setting?

Seems like you were right. As we don't have any sensitive information stored with CircleCi, I flipped ["Permissive building of fork pull requests"](https://circleci.com/gh/dlang/phobos/edit#advanced-settings) to "on" and it seems to be building now :)

While I was looking at our config I also realized that a bug in Dscanner has been fixed and we can now re-enable the style checking for `std/conv.d`. Sorry for the noise :/

**Comment by CyberShadow** [comment]
Auto-merge toggled on

**Pull request merged**